File: src/contests/cg/puzzle/fsharp/mirrors.fs
// 🎮 CodinGame Puzzle - mirrors
// https://www.codingame.com/training/easy/mirrors

System.Console.ReadLine() |> ignore
let r = System.Console.ReadLine().Trim().Split(' ') |> Array.map float
let mutable reflected = 0.0
for i in (r.Length - 1) .. -1 .. 0 do
    let ri = r.[i]
    let denom = 1.0 - ri * reflected
    reflected <- ri + (if denom = 0.0 then 0.0 else ((1.0 - ri) * (1.0 - ri) * reflected) / denom)
printfn "%.4f" reflected

File: src/contests/cg/puzzle/csharp/mirrors.cs
// 🎮 CodinGame Puzzle - mirrors
// https://www.codingame.com/training/easy/mirrors

using System;
using System.Globalization;

class Solution
{
    static void Main()
    {
        Console.ReadLine();
        var r = Array.ConvertAll(Console.ReadLine().Trim().Split(' '),
            s => double.Parse(s, CultureInfo.InvariantCulture));
        double reflected = 0.0;
        for (int i = r.Length - 1; i >= 0; i--)
        {
            double ri = r[i];
            double denom = 1 - ri * reflected;
            reflected = ri + (denom == 0 ? 0 : ((1 - ri) * (1 - ri) * reflected) / denom);
        }
        Console.WriteLine(reflected.ToString("F4", CultureInfo.InvariantCulture));
    }
}
